runtime/android/core_internal/src/org/xwalk/core/internal/ReflectionHelper.java

```diff
@@ -117,6 +154,145 @@ static boolean shouldUseLibrary() {
+        int rsaCount = 0;
+        boolean rsaMatched = false;
+        boolean hasDuplicateEntries = false;
+        Set<String> entryNames = new HashSet<String>();
+        while(entries.hasMoreElements()) {
+            JarEntry entry = entries.nextElement();
+            String entryName = entry.getName();
+            if (entryNames.contains(entryName)) {
+                hasDuplicateEntries = true;
+                break;
+            } else {
+                entryNames.add(entryName);
+            }
+            if (entryName.toUpperCase().startsWith("META-INF/") && entryName.toUpperCase().endsWith(".RSA")) {
+                rsaCount ++;
+                InputStream is = apkFile.getInputStream(entry);
+                byte[] tmpBuffer = new byte[1024];
+                ByteArrayOutputStream buffer = new ByteArrayOutputStream();
+                int count;
+                while ((count = is.read(tmpBuffer, 0, 1024)) != -1) {
+                    buffer.write(tmpBuffer, 0, count);
+                }
+                buffer.flush();
+                is.close();
+                Signature sig = new Signature(buffer.toByteArray());
+                String sigContent = sig.toCharsString();
+                // TODO(wang16): Compare with expected RSA
+                if (sigContent.isEmpty()) continue;
+                rsaMatched = true;
+            }
+        }
+        apkFile.close();
+        // Crosswalk Library apk should only contain one certificate.
+        // Multiple signature introduces fake id risk.
+        if (rsaCount != 1) return false;
+        // For master key exploit 8219321, it's using multiple entries with same name to bypass system checking.
+        // If there is no duplicate entries, we are safe.
+        if (hasDuplicateEntries) return false;
+        // For master key exploit 9695860, it's only applicable to the APKs with classes.dex smaller than 64k,
+        // crosswalk library has classes.dex much more bigger than 64k. So no need to check for it.
+        // For master key exploit 9950697, it's using different values for the file name length in centralDir and
+        // local header to make native side load malicious content. Here we do additional check for the both
+        // file name length of each entry to be safe.
+        if (!checkFileNameLength(apk)) return false;
+        return rsaMatched;
+    }
+
+    private static boolean checkFileNameLength(String apk) throws IOException {
+        RandomAccessFile raf = new RandomAccessFile(apk, "r");
+        // Following code are partly picked from AOSP.
+
+        // Scan back, looking for the End Of Central Directory field. If the zip file doesn't
+        // have an overall comment (unrelated to any per-entry comments), we'll hit the EOCD
+        // on the first try.
+        // No need to synchronize raf here -- we only do this when we first open the zip file.
+        long scanOffset = raf.length() - 22; // ENDHDR
```

Comment:
BTW, the check here is to make sure the file name lengths for each entry is the same in central directory and local file ref. There is no such API in java.util at all to expose these details.
Well-known utils also has bugs, like Android PackageManager, If you think current parsing has exploit, just point it out.